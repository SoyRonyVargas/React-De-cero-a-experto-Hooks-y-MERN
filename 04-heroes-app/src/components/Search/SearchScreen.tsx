import getHeroesByName from '../../selectors/getHeroesByName'
import { useNavigate , useLocation } from 'react-router-dom'
import { useForm } from "../../hooks/useForm"
import HeroCard from '../Hero/HeroCard'
import queryString from 'query-string'

const SearchScreen = () => {
  
  const navigate = useNavigate()
  const location = useLocation()

  const { q = '' } = queryString.parse(location.search)

  const heroes = getHeroesByName(String(q))

  const { formState , onInputChange } = useForm({
    search: q
  })

  const onSubmit = ( event : any ) => {

    event.preventDefault()

    navigate(`?q=${formState.search}`)

  }
  
  return (
    <div className="row">
      <div className="col-5">
        <h1>Search</h1>
        <hr />
        <form onSubmit={onSubmit}>
          <div className="mb-3">
            <input 
              type="text" 
              name="search"
              autoComplete="off"
              className="form-control" 
              placeholder="Search a hero..."
              value={formState.search}
              onChange={onInputChange}
            />
          </div>
          <button type="submit" className="btn btn-primary">Search</button>
        </form>
      </div>
      <div className="col-7">
        <h1>Results</h1>
        <hr />
        
        {
          heroes.length > 0 &&
          <div className="alert alert-primary" role="alert">
          Results for <b> { q } </b>
          </div>
        }
        
        {
          (heroes.length === 0 && q) &&
          <div className="alert alert-danger" role="alert">
            No results for <b> { q } </b>
          </div>
        }

        <section className=''>
          {  
            heroes.map( hero => 
              <HeroCard key={hero.id} auto={true} {...hero} />
            )
          }
        </section>          
      </div>
    </div>
  )
}

export default SearchScreen