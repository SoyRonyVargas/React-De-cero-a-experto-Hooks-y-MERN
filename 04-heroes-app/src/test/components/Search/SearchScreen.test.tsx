import SearchScreen from "../../../components/Search/SearchScreen"
import { fireEvent, render , screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"

const mockedUseNavigate = jest.fn()

jest.mock("react-router-dom", () => ({
    ...jest.requireActual('react-router-dom') as any,
    useNavigate: () => mockedUseNavigate
}));

describe('Pruebas al componente <SearchScreen/> ', () => { 

    test('Debe de mostrarse correctamente', () => { 

        const { container } = render(
            <MemoryRouter>
                <SearchScreen/>
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()

    })

    test("debe de mostrar a batman y el query string" , () => {

        const { container } = render(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchScreen/>
            </MemoryRouter>
        )

        const input__search = screen.getByRole("textbox") as HTMLInputElement
        const card__hero  = container.querySelectorAll(".card")

        expect( input__search.value ).toBe("batman")
        expect(card__hero.length).toBe(1)

    })

    test('debe de mostrar un error si no encuentra el hero', () => { 

        const { container } = render(
            <MemoryRouter initialEntries={["/search?q=batman123"]}>
                <SearchScreen/>
            </MemoryRouter>
        )

        const div__no__results  = container.querySelector(".alert.alert-danger") as HTMLDivElement
        const card__hero  = container.querySelectorAll(".card")

        expect( screen.getByText("No results for") ).toBeTruthy()
        expect( div__no__results ).toBeTruthy()
        expect(card__hero.length).toBe(0)

    })

    test("debe de llamar el navigate al buscar por el form" , () => {

        const { container } = render(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen/>
            </MemoryRouter>
        )

        const button__submit = container.querySelector(".btn.btn-primary") as HTMLButtonElement
        const input__search = screen.getByRole("textbox") as HTMLInputElement

        fireEvent.change(input__search , { 
            target: {
                value: "batman"
            }
        })

        fireEvent.click(button__submit)

        expect(mockedUseNavigate).toHaveBeenCalled()
        
        expect(mockedUseNavigate).toHaveBeenCalledTimes(1)

        // screen.debug()

    })

})

export {}