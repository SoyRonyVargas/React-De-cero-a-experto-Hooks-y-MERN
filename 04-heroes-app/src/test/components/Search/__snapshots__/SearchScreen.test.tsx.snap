// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas al componente <SearchScreen/>  Debe de mostrarse correctamente 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h1>
        Search
      </h1>
      <hr />
      <form>
        <div
          class="mb-3"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="search"
            placeholder="Search a hero..."
            type="text"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h1>
        Results
      </h1>
      <hr />
      <section
        class=""
      />
    </div>
  </div>
</div>
`;
